### PR TITLE
Assembly: Initial implementation

### PR DESCRIPTION
**Pull request by PaddleStroke**
This is a proposal for an integrated assembly workbench.
It is extremely minimal and aims to give the basis to build upon.

What it includes : 
- Add Assembly workbench.
- Implement 'Create assembly' command : this commands create assembly object (App::Part of 'type' = Assembly) in current document.
- Implement 'Insert Link' command : this commands opens a taskbox that shows you your available parts and let you add them to your currently 'Active' assembly.
- Add an empty preference page.
- Add unit test basic structure.

Please don't hesitate to comment on the proposal. For example: 
- Commands
- Code structure
- Assembly object
- Icon
...

Note : 'Create assembly' and 'Insert Link' have been made from A4 and A2+ code.
Associated issue : https://github.com/FreeCAD/FreeCAD/issues/10292

### CODE REVIEW

**Comment by sliptonic** [comment]
Added to the Integrated Assembly Workbench Project.

**Comment by chennes** [comment]
It looks like you are missing
```
PYSIDE_WRAP_RC(Assembly_QRC_SRCS Resources/Assembly.qrc)
```
at the top of your Gui CMakeLists.txt file. I think the only QRC code you have right now is for building a translations QRC.

**Comment by PaddleStroke** [comment]
@chennes thanks for the suggestion, but it still doesn't work unfortunately.

**Comment by PaddleStroke** [comment]
Where this should go ? In the Assembly/Resources/CMakeFiles.txt or the main Assembly/CMakeFiles.txt ?
```
fc_target_copy_resource(AssemblyGui
        ${CMAKE_CURRENT_BINARY_DIR}
        ${CMAKE_BINARY_DIR}/Mod/Assembly
        Assembly_rc.py)
```
When in Assembly/Resources/CMakeFiles.txt it builds but then I get an error with `import Assembly_rc` saying this module doesn't exist.

**Comment by chennes** [comment]
I'm not sure what a CMake file is doing in resources, it should be unnecessary. I'll be back at a computer in an hour or so, I can look then.

**Comment by chennes** [comment]
I added those cMake lines to Assembly/Gui/CMakeLists.txt and put in the import, and it's working for me -- so try that and if it's still not working I'll add you as a remote and send a PR.

File: src/Mod/Assembly/AssemblyGlobal.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
+/***************************************************************************
+ *   Copyright (c) 2021 Werner Mayer <wmayer[at]users.sourceforge.net>     *
```
Comment (chennes) [inline src/Mod/Assembly/AssemblyGlobal.h:2]: I'd update the copyright on this one, even though it's trivial. Please use the template from the Developer's Handbook: https://freecad.github.io/DevelopersHandbook/codeformatting/

**Comment by PaddleStroke** [comment]
Ah finally it works ! Not even sure how it got to works though. There was a visual studio file that was 'unloaded' for the assembly module. Thanks Chris for the fix !

Ok so I'll work a bit on the tests as they are failing the CI now. Fix the copyrights and make the PR ready for review.

**Comment by macdroid53** [comment]
I'm not sure if this is the right place to bring this up. I'll just dive in.

**Comment by chennes** [comment]
@PaddleStroke are you attached to these commits? The history is a little disjoint, I think they should probably be squashed -- I can do it into one commit when I merge, or you can clean things up a bit yourself if you have a different organization in mind.

**Comment by PaddleStroke** [comment]
@chennes If you can squash them all and apply black while merging that would be great.

**Comment by sliptonic** [comment]
The current state of this PR is noteworthy.  While it doesn't do much in terms of assembly (yet) it's an excellent template for workbench creation.  
It has:
- basic workbench structure
- Toolbar
- Menu
- A couple of GUI commands
- A unit test example
- Cmake configuration for installation
- Task panels and icons compiled into a resource

Basically, everything a growing boy (or girl) needs to build a workbench.

I was going to add one more commit adding the framework for an importer.  After that, I'd suggest merging it and maybe linking to it as a reference template.

**Comment by chennes** [comment]
Yeah, I just told someone on the Discord that they should use this as their base template 😄

**Comment by sliptonic** [comment]
I'll run black

**Comment by chennes** [comment]
I'll send in a PR to add it to the pre-commit -- anyone working on it should make sure they've got that set up.

File: src/Mod/Assembly/Commands.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+
+def labelName(obj):
+    if obj:
+        if obj.Name == obj.Label:
+            txt = obj.Label
+        else:
+            txt = obj.Label + " (" + obj.Name + ")"
+        return txt
+    else:
+        return None
```
[the remark this answers is not in the document]
Comment (sliptonic) [reply src/Mod/Assembly/Commands.py:71]: Yes.  Good catch.   We've started with new files for the WB so we need to make sure to update the copyright at the top if we're including code verbatim.  
In this case, let's get that code out entirely.

[the remark this answers is not in the document]
Comment (PaddleStroke) [reply src/Mod/Assembly/Commands.py:71]: I'm very sorry, I don't know well copyright/licenses so I didn't know this was in breach. I will check carefully the current code to make sure there is no more of such.

[the remark this answers is not in the document]
Comment (PaddleStroke) [reply src/Mod/Assembly/Commands.py:71]: https://github.com/FreeCAD/FreeCAD/pull/10486/files
There should be now only original content.

**Comment by sliptonic** [comment]
> the Blender Asset management system would be worth looking at

I used Blender back in the day but haven't touched it in a few years.  Can you link to a video or tutorial that illustrates the concepts?

**Comment by macdroid53** [comment]
Sure, there are quite a few.. These are a good start:

[https://youtu.be/01jcDvHpOmY?si=W18Y258wtfsUEf_m](url)

[https://youtu.be/NMpjb7kl94s?si=ZMdMmwLh1hCRMunN](url)

> > the Blender Asset management system would be worth looking at
> 
> I used Blender back in the day but haven't touched it in a few years. Can you link to a video or tutorial that illustrates the concepts?

**Comment by adrianinsaval** [comment]
> When in Assembly/Resources/CMakeFiles.txt it builds but then I get an error with `import Assembly_rc` saying this module doesn't exist.

I still get this error in my builds when starting freecad. What is supposed to generate that file?

**Comment by macdroid53** [comment]
I'm totally confused on how to make other inputs to the "Integrated Assembly" work and my comment in the Assembly forum didn't garner any comments, I'm going to link it here.
Note: I'm happy to do this right, I just don't know where/how. 
[https://forum.freecad.org/viewtopic.php?t=81311](url)

**Comment by 0penBrain** [comment]
@sliptonic @PaddleStroke Please refrain from pushing code with debug prints in it. It breaks user ability to filter what they want to be reported among logs/messages/warnings/errors.
At this time where Assy WB hardly exists, forcefully having this "Loading Assembly Workbench..." in the report view is not good.

Also now I land in this PR, it clearly breaks the contributing rule https://github.com/FreeCAD/FreeCAD/blob/0d7880a902154adc22ee24df40b443c9aa3a7d8a/CONTRIBUTING.md?plain=1#L55
If I'm correct @chennes solved this by making a squash merge. But if this is the way we want to go, we should clarify it. Basically define if it's contributor or maintainer responsibility to introduce only compilable commits.

**Comment by FlachyJoe** [comment]
@0penBrain :+1:  , ideally we should keep the PRs in draft mode until the CI tests are validated. And then they should be switched back to draft at each modification request, including squash. Only clean and safe PR should be seen merged.
But if a maintainer is possibly ready to do a part of the job (as squash merge), I'm ok. Let's not be too rigid.

**Comment by sliptonic** [comment]
@0penBrain good points.  We will do better, especially on using Draft status on PRs and squashing commits. 

Let's be a bit more lax on debug code.   This is a development branch so user concerns are not the first priority.  It's also a brand new, rapidly evolving module.  There's going to be some noise until things settle down.

**Comment by 0penBrain** [comment]
> Let's be a bit more lax on debug code. This is a development branch so user concerns are not the first priority. It's also a brand new, rapidly evolving module. There's going to be some noise until things settle down.

You know as I do that it will be lost and stay as it is.
That's not a big pain to use `FreeCAD.Console.PrintLog()`. You can even alias it if you want.

**Comment by PaddleStroke** [comment]
I'll remove the warning in my card9 branch.

**Comment by sliptonic** [comment]
> > Let's be a bit more lax on debug code. This is a development branch so user concerns are not the first priority. It's also a brand new, rapidly evolving module. There's going to be some noise until things settle down.
> 
> You know as I do that it will be lost and stay as it is. That's not a big pain to use `FreeCAD.Console.PrintLog()`. You can even alias it if you want.

I don't agree.  That kind of stuff is easy/trivial to clean up later.   Sure using PrintLog() instead of print() is easy and I recommend it.  However, I assume we are talking about other kinds of debugging logic that might be less trivial.
 
Let's be realistic, the code will mature and stabilize over time.  We don't have to be militant early on, especially if doing so slows down development.   It's just code, we can fix it.

**Comment by PaddleStroke** [comment]
You are a bit harsh in your comment. What you call fixing it are few
obscure bugs, some which were not even related to my PRs.
The one about sketcher toolbar and changing windows, the feature I
introduced has been modified many times to the point I don't understand the
last fixes. Sure I can spend one or 2 days fixing this bug that rarely
happens and is not so harmful anyway. But while sketcher is still missing
an offset tool, this is far down in my priority list.
Sketcher_Dimension had few inacurracy when it came out. And as far as I
know I fixed them all.

**Comment by sliptonic** [comment]
As far as I can tell, this PR is closed and there's nothing else actionable on it.
@0penBrain, the point of the contributing process is to make development go faster and more predictable.  As a maintainer, please try to join the merge meetings so you can see how we're dealing with issues like merge/squashing and other things in realtime.

**Comment by 0penBrain** [comment]
> As a maintainer, please try to join the merge meetings so you can see how we're dealing with issues like merge/squashing and other things in realtime.

This is very unlikely to happen because of the schedule. 😉

**Comment by sliptonic** [comment]
> > As a maintainer, please try to join the merge meetings so you can see how we're dealing with issues like merge/squashing and other things in realtime.
> 
> This is very unlikely to happen because of the schedule. 😉

We are willing to move it to get more maintainers participating.  What is a better time?  There's also a Telegram maintainers group where some discussion occurs

**Comment by 0penBrain** [comment]
> We are willing to move it to get more maintainers participating. What is a better time?

Being honest, by itself setting a weekly meeting in a busy calendar is pretty difficult. :smile:
For myself, Friday would be better, after 6PM CET/CEST ideally.

> There's also a Telegram maintainers group where some discussion occurs

I'm personally reluctant to have yet another app/account/... For IM, something where you can connect with the GH account would help.

**Pull request merged**